File: Episode12-Leaderboard/Pygame/explosion.py
import pygame
import shared

class Explosion():
	def __init__(self, images:list, centre:tuple, key:str) -> None:
		''' Class variables. key: 'sm', 'lg', 'player '''
		self.images = images				# list of 8 images
		self.centre = centre				# use for all frames					# 
		self.image = images[key][0] 		# set to first image in the sequence
		self.image = images[self.key][0] 	# set to first image in the sequence
		self.rect = self.image.get_rect()	# define rectangle from image size
		self.rect.center = self.centre		# set centre for all frames
		self.frame = 0						# no of first frame
		self.time_passed = 0				# set timer to 0
		self.frame_rate = 0.1 				# 8 images played at 1 frame per 0.1 secs = 0.8 seconds
		self.active = True

	def update(self, dt):
		self.time_passed += dt
		if self.time_passed >= self.frame_rate:					# 0.1 seconds has passed
			self.time_passed = 0								# reset timer
			self.frame += 1										# increase frame number
			if self.frame >= len(self.images[self.key]):		# check if end of list?
				self.active = False								# animation finished
			else:
				self.image = self.images[self.key][self.frame] 	# next frame
				self.rect = self.image.get_rect()				# new rectangle
				self.rect.center = self.centre					# set centre to parameter value

		return self.active

	def draw(self):
		shared.screen.blit(self.image, self.rect)				# draw current frame

File: Episode12-Leaderboard/Love2D/explosion.lua
local Class = require("lib.Class")
local Explosion = Class:derive("Explosion")
local Rectangle = require "lib.Rectangle"

function Explosion:new(images, centre, scale)
	--[[ class constructor, takes table of animation images
		 centre of the mob destroyed {x,y} , and scale factor]]
	self.images = images 	-- table of 8 images
	self.centre = centre	-- use for all frames
	self.scale = scale
	self.image = images[1] 		-- set to first image in the sequence
	self.frame = 1
	self.timePassed = 0
	self.frameRate = 0.1
	self.active = true
	self.rect = Rectangle(centre[1] - (self.image:getWidth() / 2) * self.scale,
						  centre[2] - (self.image:getHeight() / 2) * self.scale,
						  self.image:getWidth(),
						  self.image:getHeight())
end

function Explosion:update(dt)
	self.timePassed = self.timePassed + dt
	if self.timePassed >= self.frameRate then
		self.timePassed = 0
		self.frame = self.frame + 1
		if self.frame > #self.images then
			self.active = false
		else
			-- ensure image sequence is centered per frame 
			self.image = self.images[self.frame] --next frame
			self.rect:update(self.centre[1] - (self.image:getWidth() / 2) * self.scale,
							 self.centre[2] - (self.image:getHeight() / 2) * self.scale)
		end
	end
	return self.active -- when false, explosion can be set to nil
end

function Explosion:draw()
	love.graphics.draw(self.image, self.rect.x, self.rect.y, nil, self.scale, self.scale)
end

return Explosion